core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcServerManager.java

```diff
@@ -0,0 +1,278 @@
+                minionIdentity.getLocation());
+        KafkaConsumer<String, byte[]> consumer;
+        final ClassLoader currentClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            // Class-loader hack for accessing the org.apache.kafka.common.serialization.ByteArraySerializer
+            Thread.currentThread().setContextClassLoader(null);
+            consumer = new KafkaConsumer<String, byte[]>(kafkaConfig);
+        } finally {
+            Thread.currentThread().setContextClassLoader(currentClassLoader);
+        }
+        KafkaConsumerRunner kafkaConsumerRunner = new KafkaConsumerRunner(rpcModule, consumer, topicNameFactory.getName());
+        executor.execute(kafkaConsumerRunner);
+        LOG.info("started kafka consumer for module : {}", rpcModule.getId());
+        rpcModuleConsumers.put(rpcModule, kafkaConsumerRunner);
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public void unbind(RpcModule module) throws Exception {
+        if (module != null) {
+            final RpcModule<RpcRequest, RpcResponse> rpcModule = (RpcModule<RpcRequest, RpcResponse>) module;
+            registerdModules.remove(rpcModule.getId());
+            stopConsumerForModule(rpcModule);
+        }
+    }
+
+    private void stopConsumerForModule(RpcModule<RpcRequest, RpcResponse> rpcModule) {
+        KafkaConsumerRunner kafkaConsumerRunner  = rpcModuleConsumers.remove(rpcModule);
+        LOG.info("stopped kafka consumer for module : {}", rpcModule.getId());
+        kafkaConsumerRunner.shutdown();
+    }
+
+    public void destroy() {
+
+    }
+
+    public void setConfigAdmin(ConfigurationAdmin configAdmin) {
+        this.configAdmin = configAdmin;
+    }
+
+    private class KafkaConsumerRunner implements Runnable {
+
+        private final KafkaConsumer<String, byte[]> consumer;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+        private String topic;
+        private RpcModule<RpcRequest, RpcResponse> module;
+
+        public KafkaConsumerRunner(RpcModule<RpcRequest, RpcResponse> rpcModule, KafkaConsumer<String, byte[]> consumer, String topic) {
+            this.consumer = consumer;
+            this.topic = topic;
+            this.module = rpcModule;
+        }
+
+        public void shutdown() {
+            closed.set(true);
+            consumer.wakeup();
+        }
+
+        @Override
+        public void run() {
+            try {
+                consumer.subscribe(Arrays.asList(topic));
+                LOG.info("subscribed to topic {}", topic);
+                while (!closed.get()) {
+                    ConsumerRecords<String, byte[]> records = consumer.poll(Long.MAX_VALUE);
+                    for (ConsumerRecord<String, byte[]> record : records) {  
+                        RpcMessageProtos.RpcMessage rpcMessage = RpcMessageProtos.RpcMessage.parseFrom(record.value());
```

Comment:
We should have a mechanism to validate whether or not the TTL has already expired on the request, and if this is the case to avoid processing it preemptively.

We could do this by including an additional field in the message, something like `expiresAt` with `$now + ttl` and comparing that to the current time here.